plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,501 @@
+	if err != nil {
+		return nil, fmt.Errorf("unable to list servers: %v", err)
+	}
+
+	serverList, err := servers.ExtractServers(page)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract servers: %v", err)
+	}
+
+	return serverList, nil
+}
+
+func getVolumeList(provider *gophercloud.ProviderClient) (VolumeList, error) {
+
+	volume, err := openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create V2 volume client: %v", err)
+	}
+
+	page, err := volumes.List(volume, &volumes.ListOpts{AllTenants: true}).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("unable to list volumes: %v", err)
+	}
+
+	s := VolumeList{}
+	volumes.ExtractVolumesInto(page, &s)
+
+	return s, nil
+}
+
+func getStoragePools(provider *gophercloud.ProviderClient) (StoragePoolList, error) {
+
+	volume, err := openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to create V2 volume client: %v", err)
+	}
+
+	results, err := schedulerstats.List(volume, &schedulerstats.ListOpts{Detail: true}).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("unable to list storage pools: %v", err)
+	}
+
+	storagePoolList, err := schedulerstats.ExtractStoragePools(results)
+	if err != nil {
+		return nil, fmt.Errorf("unable to extract storage pools: %v", err)
+	}
+
+	return storagePoolList, nil
+
+}
+
+func gatherIdentityStatistics(acc telegraf.Accumulator, projectMap ProjectMap) {
+
+	// Ignore if any required data is missing
+	if projectMap == nil {
+		return
+	}
+
+	if len(projectMap) != 0 {
+		fields := FieldMap{
+			"projects": len(projectMap),
+		}
+		acc.AddFields("openstack_identity_total", fields, TagMap{})
+	}
+}
+
+func gatherHypervisorStatistics(acc telegraf.Accumulator, hypervisorList HypervisorList) {
+
+	// Ignore if any required data is missing
+	if hypervisorList == nil {
+		return
+	}
+
+	// Set some defaults in case we have no hypervisors yet
+	totals := IntegerFieldMap{}
+
+	for _, hypervisor := range hypervisorList {
+		// Accumulate overall statistics
+		totals["memory"] += megabytesToBytes(hypervisor.MemoryMB)
+		totals["memory_used"] += megabytesToBytes(hypervisor.MemoryMBUsed)
+		totals["running_vms"] += hypervisor.RunningVMs
+		totals["vcpus"] += hypervisor.VCPUs
+		totals["vcpus_used"] += hypervisor.VCPUsUsed
+
+		// Dump per hypervisor statistics
+		tags := TagMap{
+			"hypervisor": hypervisor.HypervisorHostname,
+		}
+		fields := FieldMap{
+			"memory":      megabytesToBytes(hypervisor.MemoryMB),
+			"memory_used": megabytesToBytes(hypervisor.MemoryMBUsed),
+			"running_vms": hypervisor.RunningVMs,
+			"vcpus":       hypervisor.VCPUs,
+			"vcpus_used":  hypervisor.VCPUsUsed,
+		}
+		acc.AddFields("openstack_hypervisor", fields, tags)
+	}
+
+	// Dump overall hypervisor statistics
+	if len(totals) != 0 {
+		acc.AddFields("openstack_hypervisor_total", totals.encode(), TagMap{})
```

Comment:
I think we should remove these totals measurements, these can be calculated at query time if needed.